More documentation on shuffleboard custom plugins

XzibitGG opened the pull request:
There didnt seem to be much documentation on custom widget, hence, added documentation on:

- Creating custom plugins
- Building and automatically deploying plugins to shuffleboard
- Creating custom data types and data classes
- Creating widgets and referencing them from their corresponding FXML file
- Settings default widgets for certain data types
- Exporting data classes/types and widgets using the custom plugin

Had to re-open the pull request because i forgot to add some changes.

Daltz333 commented:
Need a lot of formatting fixes. Review coming in a couple of hours.

Paraselenae commented:
I'll fix all the trailing whitespace errors

Daltz333 commented:
There is also indentation issues. It needs to be a 3 whitespace indent.

sciencewhiz commented:
Note that you don't have to close and create a new pull request to add new changes. You can just push them to your branch and they are automatically added to the PR.

Paraselenae commented:
3 whitespace indent for outside the code tabs or everything? @Daltz333

Paraselenae commented:
That should fix all the indentation errors and everything Prateek reviewed

Paraselenae commented:
Alright, this time it should _actually_ work

Daltz333 commented:
Are you aware of the shuffleboard wiki? https://github.com/wpilibsuite/shuffleboard/wiki

XzibitGG commented:
> Are you aware of the shuffleboard wiki? https://github.com/wpilibsuite/shuffleboard/wiki

Yeah, Incorporated the example from there and added on top of it. Want me to link it in the footer?

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,205 @@
+Creating A Widget
+=================
+Widgets allow us to view, change, and interact with data published through NetworkTables. The CameraServer, NetworkTables, and Base plugins provide the widgets to control basic
+data types (including FRC-specific data types). However, custom widgets allow us to control our custom data types we made in the previous sections or Java Objects.
```
Why is this on a separate line?

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,205 @@
+data types (including FRC-specific data types). However, custom widgets allow us to control our custom data types we made in the previous sections or Java Objects.
+
+The basic ``Widget`` interface inherits from the ``Component`` and ``Sourced`` interfaces. Both are basic building blocks towards making widgets and allows us to modify and display data.
+A good widget allows the end-user to modify the widget to suit their needs. An example could be to allow the user to control the range of the number slider, that is, its maximum and minimum or the orientation of the slider itself. The view of the widget or how it looks is defined using FXML. ``FXML`` is an XML based language that allows us to build the widget interface.
```
Put a newline separating paragraphs for distinction.

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,205 @@
+Defining a Widgets FXML
+-----------------------
+In this example, we will create 2 sliders to help us control the X and Y coordinates of our Point2D data type we created in previous sections.
+In order to create an empty, blank window for our widget, we need to create a ``Pane``. In simple terms a Pane is a "Parent" UI element that contains other "Child" UI elements, in this case, 2 sliders.
```
Newline between paragraphs for distinction

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,205 @@
+Now that we have finished creating our FXML file, we can now create a widget class. The widget class must include a ``@Description`` annotation that states the supported data types of the widget and the name of the widget.
+It also must include a ``@ParamatrizedController`` annotation that points to the FXML file containing the layout of the widget. Finally, the class must extend the ``SimpleAnnotatedWidget`` class.
+
+.. code-block:: java
```
Need both Java and C++ documentation

prateekma replied:
You can only make Shuffleboard widgets in Java -- Shuffleboard is written in Java.

Daltz333 replied:
Whoops. Ignore this then.

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,205 @@
+Binding Elements and Adding Listeners
+-------------------------------------
+Binding is a mechanism that allows JavaFX widgets to express direct relationships with NetworkTableEntries. For example, changing a widget will change its related NetworkTableEntry and vise versa.
+An example, in this case, would be changing the X and Y coordinate of our 2D point by changing slider1 and slider2 respectively.
```
Newlines between paragraphs for distinction.

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,158 @@
+-----------------------
+
+In this example, we will create a custom data type for a 2D Point and its x and y coordinates. In order to create a custom data type class,
+it must extend the abstract class `ComplexData <https://github.com/wpilibsuite/shuffleboard/blob/master/api/src/main/java/edu/wpi/first/shuffleboard/api/data/ComplexData.java>`_. Your
```
Shouldn't be on a newline

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,158 @@
+      @Override
+      public boolean equals(MyPoint2D point) {
+         // Checks if the x and y values are the same in the point being passed in and the object this method is performed on.
+         // The equals method checks to make sure the content of the objects is the same, whereas the == operator checks if they point to the same memory location.
```
Code blocks can and *should* have new lines. This comment is too long

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/index.rst:
```diff
@@ -6,4 +6,5 @@ Shuffleboard
 
    getting-started/index
    layouts-with-code/index
+   custom-widgets/index
```
This should be below advanced-usage

carbotaniuman commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,207 @@
+A good widget allows the end-user to modify the widget to suit their needs. An example could be to allow the user to control the range of the number slider, that is, its maximum and minimum or the orientation of the slider itself. The view of the widget or how it looks is defined using FXML. ``FXML`` is an XML based language that allows us to build the widget interface.
+
+Defining a Widgets FXML
+-----------------------
```
Seems out-of-scope to describe FXML in here, maybe a link to some resources?

carbotaniuman commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,207 @@
+   </StackPane>
+
+Now that we have finished creating our FXML file, we can now create a widget class. The widget class must include a ``@Description`` annotation that states the supported data types of the widget and the name of the widget.
+It also must include a ``@ParamatrizedController`` annotation that points to the FXML file containing the layout of the widget. Finally, the class must extend the ``SimpleAnnotatedWidget`` class.
```
A mention of putting the .fxml in the same package as the Java class would be helpful.

carbotaniuman commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,86 @@
+.. code-block:: groovy
+
+   dependencies {
+      api files("path/to/shuffleboard.jar")
```
```
dependencies {
    compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
    compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
}
```
This is probably the best way to source the files.

carbotaniuman commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,86 @@
+
+Deploying Plugin To Shuffleboard
+--------------------------------
+In order to load a plugin to the shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
```
Would a mention of a fat-jar with dependencies be useful for preventing a common footgun?

carbotaniuman commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,207 @@
+=================
+Widgets allow us to view, change, and interact with data published through NetworkTables. The CameraServer, NetworkTables, and Base plugins provide the widgets to control basic data types (including FRC-specific data types). However, custom widgets allow us to control our custom data types we made in the previous sections or Java Objects.
+
+The basic ``Widget`` interface inherits from the ``Component`` and ``Sourced`` interfaces. Both are basic building blocks towards making widgets and allows us to modify and display data.
```
Explanation of these interface (what they do and their role) would be helpful.

SamCarlberg commented:
> Yeah, Incorporated the example from there and added on top of it. Want me to link it in the footer?

Better to copy it into the official docs to keep everything in one place. The wiki was written when we were still on screensteps which had no good way to have community involvement.

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,156 @@
+      return new MyPoint2D(this.x, newY);
+   }
+
+This creates a new ``MyPoint2D`` object and returns it with the new y-coordinate.
```
You mix `X` and `Y` in this example. Keep it consistent, and mention that the same should be done for the other field

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,156 @@
+
+   }
+
+The class constructor is set to private to ensure that only a single instance of the data type will exist.
```
:+1:

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,156 @@
+
+.. code-block:: java
+
+   public final class MyDataType extends SimpleDataType<double> {
```
Naming should be `DoubleType` (or maybe `ExampleDoubleType`). `MyDataType` is a poor name when it's for a concrete example.

Can't do generics on primitive types; you need to use the boxed class `Double`

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,156 @@
+      private static final String NAME = "MyPoint2D";
+      public static final PointDataType Instance = new PointDataType();
+
+      public PointDataType() {
```
```suggestion
      private PointDataType() {
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,156 @@
+
+      @Override
+      public List<DataType> getDataTypes() {
+         ImmutableList.of(PointDataType.Instance);
```
This could also be `List.of` (added in Java 10, after Shuffleboard was first written)

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+--------
+Plugins provide the ability to create custom widgets, layouts, data sources/types, and custom themes. Shuffleboard provides the following plugins:
+
+- NetworkTables Plugin: To view data published on the NetworkTables
```
```suggestion
- NetworkTables Plugin: To connect to data published over NetworkTables
```

SamCarlberg replied:
Or maybe 'to interact with data...'

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+      import edu.wpi.first.shuffleboard.api.plugin.Description;
+      import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+
+      @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
```
Need to explain how these attributes are used, and the version in particular. Link to [semver.org](http://www.semver.org) to explain version numbers and how Shuffleboard treats plugin dependencies.

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+   dependencies {
+      compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
+      compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
+   }
```
Also need to define the FRC maven server in the `repositories` block

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. code-block:: java
+
+   @Requires(group = "com.example", name = "Good Plugin", minVersion = "1.2.3")
```
Need to describe what `minVersion` represents

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+
+   task deployWidget (type: Copy, group: "...", description: "...", dependsOn: "build") {
+      from "build/libs"
+      into "path/to/Shuffleboard/plugins"
```
Typically `~/Shuffleboard/plugins`, though it might end up being `~/wpilib/Shuffleboard/plugins`

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. code-block::groovy
+
+   task deployWidget (type: Copy, group: "...", description: "...", dependsOn: "build") {
```
I'd name this `deployPlugin`, as that's what's actually happening. A plugin doesn't necessarily have to provide any widgets

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+   }
+
+
+The ``deployWidget`` task takes 4 parameters, ``type: Copy`` parameter makes the task implement the `CopySpec <https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html>`_ interface
```
4 -> four

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,87 @@
+In the body, the ``from`` field specifies from where the file is to be copied from, followed by the ``into`` field specifying the destination to where the file needs to be copied.
+Finally, the ``include`` field ensures the ``.jar`` extention is also copied.
+
+By running ``gradle deployWidget`` from the command line, the jar file will automatically placed into the shuffleboard plugin folder.
```
`gradle deployPlugin`, 'shuffleboard' -> Shuffleboard'

carbotaniuman commented:
@SamCarlberg I was going to contribute a section on data recording. Is the recording function only a feature of the data source? Is there any resource on having people write code that can work with the recording functionality?

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,212 @@
+
+More about FXML can be found `here <https://docs.oracle.com/javase/8/javafx/api/javafx/fxml/doc-files/introduction_to_fxml.html>`_.
+
+Defining a Widgets FXML
```
```suggestion
Defining a Widget's FXML
```

SamCarlberg commented:
Recording is done at the data source level by calling Recorder.getInstance().record(this) when the data changes. NetworkTables has a more custom approach due to its structured nature allowing any key or table to be a data source. [link](https://github.com/wpilibsuite/shuffleboard/blob/50826b3339dce603f7396be05bb0e6c21f451ab9/plugins/networktables/src/main/java/edu/wpi/first/shuffleboard/plugin/networktables/NetworkTablesPlugin.java#L103) 

Playback is done automatically and does not require any custom logic; it's free as long as the data source is able to have its value set by code (which any valid implementation should support)

XzibitGG commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,212 @@
+
+Defining a Widgets FXML
+-----------------------
+In this example, we will create 2 sliders to help us control the X and Y coordinates of our Point2D data type we created in previous sections. It is helpful to place the FXML file in the same package as the Java class. 
```
Well noted. However, in order to create a control which allows us to input any coordinate to a 2D point would involve creating some sort of TextView control and only filtering numerical inputs in the listener which seemed a little out of scope for the tutorial. Instead, 2 number sliders would provide a better example.

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+
+      }
+
+More resources on how these attributes are used, including version numbers Shuffleboard's plugin dependencies can be found `here <https://semver.org/>`_.
```
This feels incomplete. semver.org only describes the versioning scheme.

XzibitGG replied:
The _group_, _name_, and _summary_ parameters seem quite self explanatory, versioning was specified since it must follow the systematic versioning guidelines.

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+
+More resources on how these attributes are used, including version numbers Shuffleboard's plugin dependencies can be found `here <https://semver.org/>`_.
+
+Note the ``@Description`` annotation which is needed to tell the plugin loader the properties of the custom plugin class.
```
```suggestion
Note the ``@Description`` annotation is needed to tell the plugin loader the properties of the custom plugin class.
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+More resources on how these attributes are used, including version numbers Shuffleboard's plugin dependencies can be found `here <https://semver.org/>`_.
+
+Note the ``@Description`` annotation which is needed to tell the plugin loader the properties of the custom plugin class.
+Plugin classes are permitted to have a defualt constructor but it cannot take any arguments.
```
```suggestion
Plugin classes are permitted to have a default constructor but it cannot take any arguments.
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+Building plugin
+---------------
+Plugins require the usuage of the `Shuffleboard API Library <https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/shuffleboard/api/>`_. These dependencies can be resolved in the
+``build.gradle`` file or using maven. The dependencies would be as follows:
```
```suggestion
Plugins require the usage of the `Shuffleboard API Library <https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/shuffleboard/api/>`_. These dependencies can be resolved in the ``build.gradle`` file or using maven. The dependencies would be as follows:
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+
+For Maven:
+
+ .. code-blocl:: groovy
```
No example for Maven

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+
+Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file.
+When having Plugin dependencies to other plugins, it is good practise to define those dependencies so the plugin does not load when the dependencies do not load as well.
+This can be done using the ``@Requires`` annotation as shown below:
```
```suggestion
Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file. When having Plugin dependencies to other plugins, it is good practice to define those dependencies so the plugin does not load when the dependencies do not load as well. This can be done using the ``@Requires`` annotation as shown below:
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+      include "*.jar"
+   }
+
+The path to your Shuffleboard plugin folder will most likely be ``~/Shuffleboard/plugins`` or ``~/wpilib/Shuffleboard/plugins`` .
```
I do not recommend mentioning ~/wpilib/Shuffleboard/plugins until it is changed in Shuffleboard

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,101 @@
+
+The path to your Shuffleboard plugin folder will most likely be ``~/Shuffleboard/plugins`` or ``~/wpilib/Shuffleboard/plugins`` .
+
+The ``deployPlugin`` task takes four parameters, ``type: Copy`` parameter makes the task implement the `CopySpec <https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html>`_ interface
```
It feels like there is a missing gradle example in this section

XzibitGG replied:
Are you referring to adding a complete build.gradle example?

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+Creating Custom Data Types
+==========================
+
+Widgets allow us to control and visualize different types of data. This data can be numerical in the form to integers, double, longs, etc..., or less primitive forms of data, such as Java objects, or data involving multiple fields. In order to display these types of data using widgets, it is helpful to create a container class for them.
```
`This data can be numerical in the form to integers, double, longs, etc..., or less primitive forms of data, such as Java objects, or data involving multiple fields.` This is awkwardly worded

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+==========================
+
+Widgets allow us to control and visualize different types of data. This data can be numerical in the form to integers, double, longs, etc..., or less primitive forms of data, such as Java objects, or data involving multiple fields. In order to display these types of data using widgets, it is helpful to create a container class for them.
+It is not nessecary to create your own Data Class if the widget will handle single fieled data types such as doubles, arrays, or strings.
```
```suggestion
It is not necessary to create your own Data Class if the widget will handle single fielded data types such as doubles, arrays, or strings.
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+Creating The Data Class
+-----------------------
+In this example, we will create a custom data type for a 2D Point and its x and y coordinates. In order to create a custom data type class,
+it must extend the abstract class `ComplexData <https://github.com/wpilibsuite/shuffleboard/blob/master/api/src/main/java/edu/wpi/first/shuffleboard/api/data/ComplexData.java>`_. Your
```
Do all custom data types need to extend ComplexData?

XzibitGG replied:
A custom data type should only be created if it involves multiple fields. A "SimpleData" class would simply mean that the data type only deals with one field in which case, we can use the [shuffleboard datatypes](https://github.com/wpilibsuite/shuffleboard/tree/76273d484c12a50356891b0ba7f7890b2ef25bc6/api/src/main/java/edu/wpi/first/shuffleboard/api/data/types) for whatever data type that one field is.

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+
+      @Override
+      public boolean equals(Object Object) {
+         ..
```
Should flesh out, since it probably only takes one line

XzibitGG replied:
I'll just take it out since I've already explained that 

> It is also good practise to override the default ``equals`` and ``hashcode`` methods to ensure that different objects are considered equivalent when their fields are the same.

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+    }
+
+Other methods can be added to retrieve or edit fields and instance variables, however, it is good practise to make these classes immutable to prevent changing the source data objects.
+Instead, you can make a new copy object instead of manipulating the existing object. For example, if we wanted to change the y coordinate of our point, we can define the following method:
```
```suggestion
Other methods can be added to retrieve or edit fields and instance variables, however, it is good practice to make these classes immutable to prevent changing the source data objects. Instead, you can make a new copy object instead of manipulating the existing object. For example, if we wanted to change the y coordinate of our point, we can define the following method:
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,151 @@
+
+This creates a new ``MyPoint2D`` object and returns it with the new y-coordinate. Same can be done for changing the x coordinate.
+
+Creating a Data Types
```
It feels like this section should be at the beginning of this page. 

```suggestion
Creating a Data Type
```

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+A good widget allows the end-user to customize the widget to suit their needs. An example could be to allow the user to control the range of the number slider, that is, its maximum and minimum or the orientation of the slider itself. The view of the widget or how it looks is defined using FXML. ``FXML`` is an XML based language that is useful for defining the static layout of the widget (Panes, L ...
+
+More about FXML can be found `here <https://docs.oracle.com/javase/8/javafx/api/javafx/fxml/doc-files/introduction_to_fxml.html>`_.
```
Is there a Java 11 version of this page?

sciencewhiz commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+Layout panes are also extremely useful for placing child nodes in one horizontal row using a `HBox <https://docs.oracle.com/javase/8/javafx/api/javafx/scene/layout/HBox.html>`_ or one vertical column using a `VBox <https://docs.oracle.com/javase/8/javafx/api/javafx/scene/layout/VBox.html>`_.
+
+The basic syntax for defining a Pane uing FXML would be as the following:
```
```suggestion
The basic syntax for defining a Pane using FXML would be as the following:
```

XzibitGG commented:
Examples were used from the shuffleboard wiki, and are in the [shuffleboard repo ](github.com/wpilibsuite/shuffleboard/tree/master/example-plugins/custom-data-and-widget)

AustinShalit commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -1,6 +1,6 @@
 Creating Custom Data Types
 ==========================
-Widgets allow us to control and visualize different types of data. This data can be numerical in the form to integers, double, longs, etc..., or less primitive forms of data, such as Java objects, or data involving multiple fields. In order to display these types of data using widgets, it is helpful to create a container class for them.
+Widgets allow us to control and visualize different types of data. This data could be integers and doubles or even Java Objects.. In order to display these types of data using widgets, it is helpful to create a container class for them.
```
```suggestion
Widgets allow us to control and visualize different types of data. This data could be integers and doubles or even Java Objects. In order to display these types of data using widgets, it is helpful to create a container class for them.
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,143 @@
+
+   public final class MyDoubleDataType extends SimpleDataType<Double> {
+
+      private static final String NAME = "Douoble";
```
```suggestion
      private static final String NAME = "Double";
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
+
+   /*If the FXML file and Java file are in the same package, that is the java file is in src/main/java and the
```
`/*` should be on its own line

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+.. code-block:: java
+
+   xSlider.valueProperty().addListener((observable, oldValue, newValue) -> setData(newValue));
```
Since the data type is `Point2D`, `setData(double)` won't work

```suggestion
   xSlider.valueProperty().addListener((observable, oldValue, newValue) -> setData(getData().withX(newValue));
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+.. code-block:: xml
+
+   <?import javafx.scene.layout.*?>
+   <GridPane xmlns:fx="http://javafx.com/fxml/1" fx:controller="/path/to/widget/class" fx:id="root">
```
Why GridPane here? StackPane would be better given that there's only one direct child node

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+.. code-block:: xml
+
+   <?import javafx.scene.layout.*?>
+   <GridPane xmlns:fx="http://javafx.com/fxml/1" fx:controller="/path/to/widget/class" fx:id="root">
```
A brief explanation of what the `fx:controller` bit does might be useful. The JavaFX docs presumably go more in depth, but having just one or two sentences here would be nice

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+      private Slider ySlider;
+   }
+
+In order to display our pane on our custom widget we need to override the ``getView()`` method and return our ``StackedPane``.
```
```suggestion
In order to display our pane on our custom widget we need to override the ``getView()`` method and return our ``StackPane``.
```

SamCarlberg replied:
Also, the docs here say `StackPane` but the FXML uses `GridPane`. Pick one and stick to it

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+-------------------------------------
+Binding is a mechanism that allows JavaFX widgets to express direct relationships with the data source. For example, changing a widget will change its related NetworkTableEntry and vise versa.
+
+An example, in this case, would be changing the X and Y coordinate of our 2D point by changing out xSlider and ySlider respectively.
```
"changing out"? Not sure what you're trying to say here

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+An example, in this case, would be changing the X and Y coordinate of our 2D point by changing out xSlider and ySlider respectively.
+
+A good practice is to set bindings in the ``initalize()`` method tagged with the ``@FXML`` annotation which will be called by the FXML framework.
```
The annotation is actually required to be present if the method is non-`public`

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import javafx.fxml.FXML;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
```
Import ordering

```suggestion
   import edu.wpi.first.shuffleboard.api.widget.Description;
   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
   import javafx.fxml.FXML;
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import javafx.fxml.FXML;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
```
```suggestion
   import edu.wpi.first.shuffleboard.api.widget.Description;
   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
   import javafx.fxml.FXML;
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import javafx.fxml.FXML;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
```
```suggestion
   import edu.wpi.first.shuffleboard.api.widget.Description;
   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
   import javafx.fxml.FXML;
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
```
```suggestion
   import edu.wpi.first.shuffleboard.api.widget.Description;
   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+   @Override
+   public Map<DataType, ComponentType> getDefaultComponents() {
+      return Map.of(new Point2DType(), WidgetType.forAnnotatedWidget(Point2DWidget.class));
```
```suggestion
      return Map.of(Point2DType.Instance, WidgetType.forAnnotatedWidget(Point2DWidget.class));
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+Set Default Widget For Data type
+--------------------------------
+In order to set your widget as default for your custom data type, you can overide the ``getDefaultComponents()`` in your plugin class that stores a Map for all default widgets as noted below:
```
```suggestion
In order to set your widget as default for your custom data type, you can override the ``getDefaultComponents()`` in your plugin class that stores a Map for all default widgets as noted below:
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst:
```diff
@@ -0,0 +1,199 @@
+
+In this case, the ``setData()`` method sets the value in the data source of the widget to the ``newValue``.
+
+
```
Need to document `getComponents()`. All custom components need to be exported (including ones that are the default for some data type):

```java
  @Override
  public List<ComponentType> getComponents() {
    return List.of(WidgetType.forAnnotatedWidget(Point2DWidget.class));
  }
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,143 @@
+
+Creating a Data Type
+---------------------
+There are two different data types that can be made, Simple data types involve data that contains only one field (ie. a single number or string) whereas Complex data types require multiple data fields (ie. multiple strings, multiple numbers).
```
```suggestion
There are two different data types that can be made: Simple data types that have only one field (ie. a single number or string), and Complex data types that have multiple data fields (ie. multiple strings, multiple numbers).
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,143 @@
+      }
+
+      @Override
+      public double getDefaultValue() {
```
Need to return the boxed type, unfortunately.

```suggestion
      public Double getDefaultValue() {
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst:
```diff
@@ -0,0 +1,143 @@
+
+Exporting Data Type To Plugin
+-----------------------------
+In order to have the data type be recognized by the Shuffleboard, the plugin must export them by overriding the ``getDataTypes`` method.
```
Just "Shuffleboard", not "the Shuffleboard"

```suggestion
In order to have the data type be recognized by Shuffleboard, the plugin must export them by overriding the ``getDataTypes`` method.
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,95 @@
+
+Deploying Plugin To Shuffleboard
+--------------------------------
+In order to load a plugin to the shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
```
```suggestion
In order to load a plugin in Shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
```

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,95 @@
+
+By running ``gradle deployPlugin`` from the command line, the jar file will automatically placed into the Shuffleboard plugin folder.
+
+Manually Adding Plugin
```
Plugins added through the UI have their paths cached. The "Clear cache" button clears that cache so users can fix issues or just drop plugins they don't use any more.

XzibitGG commented:
Thanks sam! Really appreciate you taking the time to review this pr.

SamCarlberg commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -83,6 +83,8 @@ specifying what to copy. The group and description parameters to specifiy what t
 In the body, the ``from`` field specifies from where the file is to be copied from, followed by the ``into`` field specifying the destination to where the file needs to be copied.
 Finally, the ``include`` field ensures the ``.jar`` extention is also copied.
 
+After deploying to Shuffleboard, open Shuffleboard, and use Ctrl + L to open the plugins menu, and select clear cache. This is necessary since Shuffleboard automatically caches the plugin when it is loaded, and clearing the cache updates the cached version to the file in the plugins folder.
```
The cache is just so Shuffleboard knows where the plugin JAR is so it can be loaded automatically at the next startup. It doesn't cache specific versions. See [here](https://github.com/wpilibsuite/shuffleboard/blob/60076f619a67a4a43bb7f7854123e1bedf212bcf/app/src/main/java/edu/wpi/first/shuffleboard/app/PluginPaneController.java#L162) and [here](https://github.com/wpilibsuite/shuffleboard/blob/60076f619a67a4a43bb7f7854123e1bedf212bcf/app/src/main/java/edu/wpi/first/shuffleboard/app/plugin/PluginCache.java)

Daltz333 commented on source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst:
```diff
@@ -0,0 +1,97 @@
+In the body, the ``from`` field specifies from where the file is to be copied from, followed by the ``into`` field specifying the destination to where the file needs to be copied to.
+Finally, the ``include`` field ensures all files with the ``.jar`` extention is also copied.
+
+After deploying, Shuffleboard will cache the path of the plugin so it can be automatically loaded the next time Shuffleboard loads. It may be necessary to click on ``Clear Cache`` under the plugins menu to remove a plugin or fix issues in the plugin before loading it again.
```
"Fix issues in the plugin", what does this entail?

carbotaniuman replied:
It's the equivalent of restarting the app but for the plugin system, ie picking up an old version or just acting weird.
